org.lflang/src/org/lflang/generator/ReactionInstanceGraph.java

```diff
@@ -254,4 +277,43 @@ private void assignLevels() {
             removeNode(origin);
         }
     }
+    
+    /**
+     * Assign a level to a reaction runtime instance.
+     * 
+     * @param runtime The reaction runtime instance.
+     * @param level The level to assign.
+     */
+    private void assignLevel(Runtime runtime, Integer level) {
+        runtime.level = level;
+        adjustMaxNumOfReactionPerLevel(level, 1);
+    }
+    
+    /**
+     * Update the level of the reaction <code>runtime<code> instance
+     * to <code>level<code> if <code>level<code> is larger than the
+     * level already assigned to <code>runtime<code>.
+     */
+    private void updateLevel(Runtime runtime, Integer level) {
+        if (runtime.level < level) {
+            runtime.level = level;
+            // Adjust the <code>maxNumOfReactionPerLevel<code> field
+            // accordingly.
+            adjustMaxNumOfReactionPerLevel(runtime.level, -1);
+            adjustMaxNumOfReactionPerLevel(level, 1);
+        }
+    }
+    
+    /**
+     * Adjust {@link #maxNumOfReactionPerLevel} at index <code>level<code> by
+     * <code>valueToAdd<code>.
+     */
+    private void adjustMaxNumOfReactionPerLevel(int level, int valueToAdd) {
+        if (maxNumOfReactionPerLevel.size() > level) {
+            maxNumOfReactionPerLevel.set(level, maxNumOfReactionPerLevel.get(0) + valueToAdd);
```

Comment:
This I don't understand. Why is this happening?

Reply:
This is adjusting the number of reactions for `level` by the requested `valueToAdd`.

But, `maxNumOfReactionPerLevel` is a list and it might not have an element at index `level`, hence the `else` statement.

Reply:
I think I see at least one bug in `updateLevel`. Let's wait for @edwardalee.